Adds middleware for persistent cookies

TheLonelyGhost opened the pull request:
Not much to say. This is some middleware to allow cookies to persist between HTTP requests. Currently, it doesn't honor a lot of the constraints cookies require (e.g., path, expiration, http-only), but it does allow for tracking a persistent session.

TheLonelyGhost commented:
Note that Crystal (as of v0.18.7) has a bug with the following snippet, where it can't find `config` for the phrase `config.cookies`:

```crystal
client = Cossack::Client.new("http://example.com") do |config|
  config.use Cossack::CookieJarMiddleware, cookie_jar: config.cookies
end
```

A workaround is to write like this:

```crystal
client = Cossack::Client.new("http://example.com")
client.use Cossack::CookieJarMiddleware, cookie_jar: config.cookies
```

greyblake commented:
Thanks! I'll take a look after work)

greyblake commented on src/cossack/cookie_jar.cr:
```diff
@@ -0,0 +1,46 @@
+      end
+    end
+
+    def self.from_file(file_path : String)
```
Could you please move the class method to the top?

TheLonelyGhost replied:
Only reason I kept it here was for close proximity to the instance-method version. They used to be named the same but then I thought better of it.

Will do.

greyblake commented on src/cossack/cookie_jar.cr:
```diff
@@ -0,0 +1,46 @@
+      headers
+    end
+
+    def add_response_headers(headers)
```
The reason you decide to override those methods, because they delete `Cookie`, `Set-Cookie` headers?

TheLonelyGhost replied:
I overrode this because if there were no cookies, it would set an http header `Cookie` (for example) with a value of an empty string.

This could have been in a pull request for the Crystal standard library, but putting it here allows it to be used immediately.

greyblake commented on spec/acceptance/cookie_jar_middleware_spec.cr:
```diff
@@ -0,0 +1,36 @@
+    expect(response.headers["Set-Cookie"]?).not_to be_nil
+    expect(client.cookies.empty?).to be_false
+
+    expect(client.get("/cookie").body).to match(/^Cookie: /)
```
Could you please here check presence of actual values (`test_cookie=works`)

greyblake commented on spec/acceptance/cookie_jar_middleware_spec.cr:
```diff
@@ -0,0 +1,36 @@
+    client.cookies << HTTP::Cookie.new("some_cookie", "some_value")
+
+    expect(client.cookies.empty?).to be_false
+    expect(client.get("/cookie").body).to match(/^Cookie: /)
```
Can you please check value of cookies? (`some_cookie=some_value`)

greyblake commented:
@TheLonelyGhost Thanks again for the nice pull request! I guess the feature is awesome and really needed.
I left some minor notes, can you please address them?
Also, if it's not hard for you, can you please add note about `CookieJarMiddleware` to  README (how to use it and how to import/export cookies from/to files) and add yourself to contributors? =)

Thanks!

TheLonelyGhost commented:
I've added in tests for cookie value, restructured so the class methods are at the top of the class definition, and added a lot of documentation on how to use it and why certain decisions were made. I then rebased the code from the newly update master branch (thanks to #3 being merged).

Any other comments or changes?

greyblake commented:
@TheLonelyGhost Big thanks for the addressing the concerns:) I'm gonna merge it and bump the version.
